docs: ADR-006 Oracle (#249)
umee-network/umee: docs/architecture/ADR-006-oracle.md

## Decision

We'd like to use the concepts introduced in [Terra's Oracle](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#concepts), but with a few modifications :

- Combine the `price-feeder` and `feeder` into a single binary [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)
- Only support MsgAggregateExchangeRate(Pre)Vote, i.e. not allow individual price updates [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)
- Skip the logic for the [Cross Exchange Rate](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#compute-cross-exchange-rate-using-reference-terra), and record simplified exchange rates like `ATOMUSDT`
- Remove support for the [Tobin Tax](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-market.html#tobintax), a Terra-specific fee for when users spot trade.

## Detailed Design

- Terra's design for the voting procedure as documented [here](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#voting-procedure)

> 

Reviewer: ```suggestion

## API

The `x/oracle` module will provide the following methods on its keeper, to be used by `x/leverage`:

```go
    // example - we should discuss if this is the best way to do things
    GetValue(asset sdk.Coin) (sdk.Dec, error) // get the USD value of an input sdk.Coin
    GetTotalValue(asset sdk.Coin) (sdk.Dec, error) // return the total USD value of all coins in an sdk.Coins
```
```

- Combine the `price-feeder` and `feeder` into a single binary [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)
- Only support `MsgAggregateExchangeRate(Pre)Vote`, i.e. not allow individual price updates [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)
- Skip the logic for the [Cross Exchange Rate](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#compute-cross-exchange-rate-using-reference-terra), and record simplified exchange rates like `ATOM`
- We'll store multiple exchange rates with a base of USD, [instead of a single coin's exchange rate](https://github.com/terra-money/core/blob/746a15f1bd83d62cd284e4af9471dc58701b3e33/x/oracle/keeper/keeper.go#L88)
- Remove support for the [Tobin Tax](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-market.html#tobintax), a Terra-specific fee for when users spot trade.

## Detailed Design

- Terra's design for the voting procedure as documented [here](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#voting-procedure)

> ### API

Reviewer 2: ```suggestion
### API

```

The `x/oracle` module will provide the following method on its keeper, to be used by `x/leverage`:
```go
    GetExchangeRate(base string) (sdk.Dec, error) // get the USD value of an input base denomination
```

## Consequences

### Positive

- More control over how our oracle works, so that we can integrate it with the `x/leverage` module how we like
- Less overhead
- Cleaner code

### Negative

- More time spent on `x/oracle` development

> ### Neutral

Reviewer 2: ```suggestion
### Neutral

```

# ADR 006: Oracle

## Changelog

- November 29th, 2021: Initial Draft (@adamewozniak)

## Status

Proposed

## Context

> Umee needs an oracle to determine the price of assets.
> 
> From section 5.1 of the [Umee Whitepaper](https://umee.cc/umee-whitepaper/):

Reviewer 2: ```suggestion
Umee needs an oracle to determine the price of assets. From section 5.1 of the [Umee Whitepaper](https://umee.cc/umee-whitepaper/):
```

> Oracle reads asset price and updates the state to the Asset Facility Coordination
Protocol

> We've already decided to import a large chunk of this from [Terra's oracle](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html), although a few parts of this are specific to Terra's protocol and do not need to be implemented.

Reviewer 2: ```suggestion
We've already decided to import a large chunk of this from [Terra's oracle](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html), although a few parts of this are specific to Terra's protocol and do not need to be implemented with respect to Umee.
```

## Alternative Approaches

- Cloning the x/oracle module completely. This would leave our code dirty, and we'd later have issues interfacing with Terra's [Cross Exchange Rate](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#compute-cross-exchange-rate-using-reference-terra), since it's designed for getting the exchange rate of only Terra.

> - Using something like Band or Chainlink. This would be additional overhead, and we'd [have less control](https://github.com/umee-network/umee/issues/97#issuecomment-923914840) over how our oracle works.

Reviewer 2: I still don't know or understand how we'd use an external Oracle service like Band or Chainlink. Do you know how?

Author: I'm not sure about band, but I believe Chainlink is specifically built for smart contracts, so I don't think we could even use it for validator modules - [chainlink api](https://docs.chain.link/docs/request-and-receive-data/)

Author: Looks like [somebody did it](https://betterprogramming.pub/connect-a-chainlink-oracle-to-a-cosmos-blockchain-d7934d75bae5) using chainlink's [external initiators](https://docs.chain.link/docs/external-initiators-introduction/), although this seems more convoluted than building our own for what we specifically need

Author: Added a couple references to that implementation in the adr

## Decision

We'd like to use the concepts introduced in [Terra's Oracle](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#concepts), but with a few modifications :

- Combine the `price-feeder` and `feeder` into a single binary [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)

> - Only support MsgAggregateExchangeRate(Pre)Vote, i.e. not allow individual price updates [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)

Reviewer 2: ```suggestion
- Only support `MsgAggregateExchangeRate(Pre)Vote`, i.e. not allow individual price updates [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)
```

- Skip the logic for the [Cross Exchange Rate](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#compute-cross-exchange-rate-using-reference-terra), and record simplified exchange rates like `ATOM`
- We'll store multiple exchange rates with a base of USD, [instead of a single coin's exchange rate](https://github.com/terra-money/core/blob/746a15f1bd83d62cd284e4af9471dc58701b3e33/x/oracle/keeper/keeper.go#L88)
- Remove support for the [Tobin Tax](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-market.html#tobintax), a Terra-specific fee for when users spot trade.

## Detailed Design

- Terra's design for the voting procedure as documented [here](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#voting-procedure)

> ### API
> The `x/oracle` module will provide the following methods on its keeper, to be used by `x/leverage`:

Reviewer 2: ```suggestion
### API

The `x/oracle` module will provide the following methods on its keeper, to be used by `x/leverage` and any other module requiring prices of assets:
```

```go

>     GetValue(asset sdk.Coin) (sdk.Dec, error) // get the USD value of an input sdk.Coin

Reviewer 2: no need to provide a coin here since the `Amount` is irrelevant ;-)

```suggestion
    GetValue(base string) (sdk.Dec, error) // get the USD value of an input base denomination
```

Reviewer 2: Also, what do you think about `GetPrice` or `GetExchangeRate` (assuming USD quote) instead?

Reviewer: Those two comments are related - I was hoping the oracle could multiply by the amount (and compute and `atom` to `uatom` etc) as part of this function, thus returning the USD value of the whole sdk.Coin

>     GetTotalValue(asset sdk.Coin) (sdk.Dec, error) // return the total USD value of all coins in an sdk.Coins

Reviewer 2: I don't understand the purpose of this one. Did you mean to use `sdk.Coins`, i.e. plural? If so, use `bases []string`

Reviewer: That plural was intentional - see my other comments, but ideally I just pass in sdk.Coin(s) and get the total USD value of what was put in, including multiplying by each denom's amount.

- Only support MsgAggregateExchangeRate(Pre)Vote, i.e. not allow individual price updates [Ref](https://github.com/umee-network/umee/issues/97#issuecomment-939610302)
- Skip the logic for the [Cross Exchange Rate](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#compute-cross-exchange-rate-using-reference-terra), and record simplified exchange rates like `ATOM`
  -  We'll store multiple exchange rates with a base of USD, [instead of a single coin's exchange rate](https://github.com/terra-money/core/blob/746a15f1bd83d62cd284e4af9471dc58701b3e33/x/oracle/keeper/keeper.go#L88)
- Remove support for the [Tobin Tax](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-market.html#tobintax), a Terra-specific fee for when users spot trade.

## Detailed Design

- Terra's design for the voting procedure as documented [here](https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#voting-procedure)

### API
The `x/oracle` module will provide the following methods on its keeper, to be used by `x/leverage`:

```go

>     // example - we should discuss if this is the best way to do things

Reviewer: ```suggestion
```
